### PR TITLE
Welling/add pii to assay type

### PR DESCRIPTION
**Pull request by jswelling**
This PR adds two new boolean properties, vis_only and contains_pii, to the assay_type data structure and assigns values for those fields to the existing assays.  

* vis_only is true if the data type contains only visualization information, and thus should be subject to vis lifting.  It defaults to False.  It exists to support future extensions to vis lifting, and because other changes make the presence of "image pyramid" in the data type list an unreliable indicator.
*  contains_pii is true if the data type contains personally identifying information and thus must be protected.  It defaults to True for safety.  

The AssayType class and the entries in assay_types.yaml are extended to support these new properties.

### CODE REVIEW

File: src/app.py
Comment (mccalluc) [inline src/app.py:77]: Delete, or change to an `info` or `debug` log?

File: src/libs/assay_type.py
Comment (mccalluc) [inline src/libs/assay_type.py:89]:
```suggestion
        self.contains_pii = this_def.get('contains-pii', True)  # default to True for safety
        self.vis_only = this_def.get('vis-only', False)  # False is more common
```
more concise

File: src/libs/assay_type.py
Comment (mccalluc) [inline src/libs/assay_type.py:129]: This testing approach is getting awkward. It essentially splits the assertions between two files: this one, and the output that you diff. Problems noticed in the diff would need to be found here and corrected. Suggest using doctests or pytest.

File: src/libs/assay_type.py
Comment (jswelling) [reply src/libs/assay_type.py:89]: Agreed; much better.

File: src/app.py
Comment (jswelling) [reply src/app.py:77]: Yeah, delete.

File: src/libs/assay_type.py
Comment (jswelling) [reply src/libs/assay_type.py]: I've converted it to doctest.  I'd thought that doing so would produce something that was unacceptably verbose, but it's not really that bad, and as you say the unification of the test prototype and the code itself has merit.

File: src/libs/assay_type.py
Comment (mccalluc) [inline src/libs/assay_type.py:34]:
```suggestion
    >>> AssayType('codex')  # should be uppercase
```
With the extra character, I'm guessing this wasn't being seen by doctest.

File: src/libs/assay_type.py
Comment (mccalluc) [inline src/libs/assay_type.py:74]: Helper function makes it easier to focus on what matters:
```suggestion
    >>> def name_desc_vis_pii(assay):
    ...     a_t = AssayType(assay)
    ...     return [a_t.name, a_t.description, a_t.vis_only, a_t.contains_pii]

    >>> name_desc_vis_pii('CODEX')
    ['CODEX', 'CODEX', False, False]

    >>> name_desc_vis_pii('codex_cytokit')  # this one does not contain PII
    ['codex_cytokit', 'CODEX [Cytokit + SPRM]', False, False]

    >>> name_desc_vis_pii('salmon_rnaseq_bulk')
    ['salmon_rnaseq_bulk', 'Bulk RNA-seq [Salmon]', False, False]

    >>> name_desc_vis_pii('scRNA-Seq-10x')  # This one does contain PII
    ['scRNA-Seq-10x', 'scRNA-seq (10x Genomics)', False, True]

    >>> name_desc_vis_pii(['PAS', 'Image Pyramid'])  # complex alt name
    ['PAS_pyramid', 'PAS Stained Microscopy [Image Pyramid]', True, False]

    >>> name_desc_vis_pii(['Image Pyramid', 'PAS'])  # complex alt name
    ['PAS_pyramid', 'PAS Stained Microscopy [Image Pyramid]', True, False]

    >>> name_desc_vis_pii('image_pyramid')  # simple vis-only alt name
    ['image_pyramid', 'Image Pyramid', True, False]

    >>> name_desc_vis_pii(['Image Pyramid', 'foo'])  # invalid complex alt name
    Traceback (most recent call last):
    ...
    RuntimeError: No such assay_type ['Image Pyramid', 'foo'], even as alternate name

    >>> name_desc_vis_pii('salmon_rnaseq_10x_sn')  # simple valid alt name
    ['salmon_sn_rnaseq_10x', 'snRNA-seq [Salmon]', False, False]
```

File: src/libs/assay_type.py
Comment (mccalluc) [inline src/libs/assay_type.py:166]:
```suggestion
        >>> assert all(elt in all_names for elt in primary_names)
        >>> assert not any(elt in nonprimary_names for elt in primary_names)
        >>> assert all(elt in all_names for elt in nonprimary_names)
        >>> assert not any(elt in primary_names for elt in nonprimary_names)
        >>> assert len(all_names) == len(primary_names) + len(nonprimary_names)
```
Your call.

File: src/libs/test.sh
Comment (mccalluc) [inline src/libs/test.sh:14]:
```suggestion
python ./assay_type.py
```
I think the failure message from doctest is sufficient: Not sure if wrapping it adds value.

**Review by mccalluc — APPROVED** [review]
Noted some places where it could be more concise, but these are style tweaks: Your call.

File: src/libs/assay_type.py
Comment (jswelling) [reply src/libs/assay_type.py:166]: Ooh, I didn't know you could put an assert in a doctest.  Yes, this is better.

File: src/libs/test.sh
Comment (jswelling) [reply src/libs/test.sh:14]: I'm going to claim that it does add value- I know it does for me, because I'm not used to doctest, and I think it would for Bill.  So I'll leave this one as it is.

**Comment by jswelling** [comment]
This closes https://github.com/hubmapconsortium/ingest-pipeline/issues/317

**Review by yuanzhou — APPROVED** [review]

**Review by shirey — APPROVED** [review]

**Comment by mccalluc** [comment]
Is the intended use of `contains_pii` to set the visibility of datasets at creation time? There is no call for the portal to use this information at run time in the same way that it might use `vis_only`, I hope?

**Review by mruffalo — APPROVED** [review]

**Comment by jswelling** [comment]
@mccalluc as far as I know there is no change to the Portal proposed in response to the contains_pii attribute.  My understanding is that the entity database infrastructure will check this attribute for consistency with its current contains-human-sequence-data flag, and that the Portal bases visibility decisions on that flag, so the Portal should not need to change its behavior.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (pdblood) [inline src/search-schema/data/definitions/enums/assay_types.yaml:23]: This is a bad key name, as it represents derived data but it is nearly identical to the raw that that DOES contain PII. The key should have some indication that this is derived data after processing.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (pdblood) [inline src/search-schema/data/definitions/enums/assay_types.yaml:226]: @jswelling why don't the lab processed datasets have a `contains-pii` property?

Comment (pdblood) [reply src/search-schema/data/definitions/enums/assay_types.yaml:226]: same goes for the one below...

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (pdblood) [inline src/search-schema/data/definitions/enums/assay_types.yaml:207]: I don't like this key for the same reason: no indication in the key that it is not raw atac-seq data, which should be protected.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (pdblood) [inline src/search-schema/data/definitions/enums/assay_types.yaml:272]: Same complaint as above about this key name not distinguishing from raw protected data...

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (pdblood) [inline src/search-schema/data/definitions/enums/assay_types.yaml:321]: Same complaint about key name as above...

**Review by pdblood — COMMENTED** [review]
@jswelling: I think all the PII assignments are correct. The only one I would want to double check out of abundance of caution is DART-FISH... I have a few other comments that you can choose to address now or later.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (ropelews) [inline src/search-schema/data/definitions/enums/assay_types.yaml:63]: Why do some of these contain vis-only field and others do not? assuming types marked vis-only=true are ONLY used for visualization.

**Review by ropelews — APPROVED** [review]
I think that these are all correctly tagged for pii.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (jswelling) [reply src/search-schema/data/definitions/enums/assay_types.yaml:226]: The lab processed datasets don't have it because I wasn't sure if we could trust them to *not* contain PII, so I left it to default to 'does contain PII'.  If we trust them to not include PII in the lab processed results, or if we do plan to actively scan for PII, we could set 'contains-pii' to False for the lab data.  Shall I do so?

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (jswelling) [reply src/search-schema/data/definitions/enums/assay_types.yaml:207]: The 'primary: true' flag is supposed to distinguish primary datasets.  I agree that a lot of the enums (like sc_atac_seq_snare in this case) are poorly chosen.  Once we have this mechanism in place we can relegate those to being 'alt-names', with better enums for the primary keys.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (jswelling) [reply src/search-schema/data/definitions/enums/assay_types.yaml:272]: That one is primary: true, meaning that it is raw data, and contains-pii: true, meaning that it does need to be protected.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (jswelling) [reply src/search-schema/data/definitions/enums/assay_types.yaml:63]: That's correct @ropelews .  The vis-ony key defaults to false, just as the contains-pii key defaults to true.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (pdblood) [reply src/search-schema/data/definitions/enums/assay_types.yaml:226]: There is a lab processed seqfish that **does** have the explicit `contains-pii: false` label, so I was confused that some lab processed data had it and some didn't. I think we can set these lab processed data to false, and then implement our checking to be extra careful for future data sets. For the lab processed data we have, I think I already verified with them that there was no protected data.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (pdblood) [reply src/search-schema/data/definitions/enums/assay_types.yaml:207]: OK, sounds good.

File: src/search-schema/data/definitions/enums/assay_types.yaml
Comment (pdblood) [reply src/search-schema/data/definitions/enums/assay_types.yaml:272]: I'm referring to `sc_atac_seq_sci` which is `primary:false`. But, I agree that we can fix these keys later.

**Comment by jswelling** [comment]
I will verify that DART-FISH does not contain PII by looking at a sample, and I will add contains-pii:false to the lab-derived data types.  After that, I will merge.

**Comment by jswelling** [comment]
Our two current DART-FISH samples contain no fastq or BAM files.

**Pull request merged**
